app/src/main/java/io/gnosis/safe/ui/settings/owner/ledger/LedgerController.kt

```diff
@@ -0,0 +1,272 @@
+    }
+
+    fun scanResults() =
+        scanResultFlow.conflate()
+
+    private val scanSettings = ScanSettings.Builder()
+        .setScanMode(ScanSettings.SCAN_MODE_LOW_POWER)
+        .build()
+
+    private val scanFilter = ScanFilter.Builder()
+        .setServiceUuid(ParcelUuid(LEDGER_SERVICE_DATA_UUID))
+        .build()
+
+    private var isScanning = false
+
+
+    fun startBleScan(fragment: Fragment, missingLocationPermissionHandler: () -> Unit) {
+        if (!bluetoothAdapter.isEnabled) {
+            promptEnableBluetooth(fragment)
+        } else {
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && !context.hasPermission(Manifest.permission.ACCESS_FINE_LOCATION)) {
+                missingLocationPermissionHandler.invoke()
+            } else {
+                if (isScanning) {
+                    stopBleScan()
+                }
+                bleScanner.startScan(listOf(scanFilter), scanSettings, scanCallback)
+                isScanning = true
+            }
+        }
+    }
+
+    fun stopBleScan() {
+        bleScanner.stopScan(scanCallback)
+        isScanning = false
+    }
+
+    fun connectToDevice(context: Context, device: BluetoothDevice, callback: DeviceConnectedCallback) {
+        if (isScanning) {
+            stopBleScan()
+        }
+        Timber.d("Attempting to connect to ${device.name}")
+        deviceConnectedCallback = callback
+        ConnectionManager.connect(device, context)
+    }
+
+    fun teardownConnection() {
+        ConnectionManager.unregisterListener(connectionEventListener)
+        connectedDevice?.let { ConnectionManager.teardownConnection(it) }
+    }
+
+    fun getAddressCommand(path: String, displayVerificationDialog: Boolean = false, chainCode: Boolean = false): ByteArray {
+
+        val paths = splitPath(path)
+
+        val commandData = mutableListOf<Byte>()
+
+        val pathsData = ByteArray(1 + paths.size)
+        pathsData[0] = paths.size.toByte()
+
+        paths.forEachIndexed { index, element ->
+            pathsData[1 + index] = element
+        }
+
+        commandData.add(0xe0.toByte())
+        commandData.add(0x02.toByte())
+        commandData.add((if (displayVerificationDialog) 0x01.toByte() else 0x00.toByte()))
+        commandData.add((if (chainCode) 0x01.toByte() else 0x00.toByte()))
+        commandData.addAll(pathsData.toList())
+
+        val command = commandData.toByteArray()
+        Timber.d(command.toHexString())
+
+        return command
+    }
+
+    suspend fun getAddress(device: BluetoothDevice, path: String): Solidity.Address = suspendCoroutine {
+        ConnectionManager.writeCharacteristic(device, writeCharacteristic!!, wrapADPU(getAddressCommand(path)))
+        addressContinuation = it
+    }
+
+    suspend fun addressesForPage(derivationPath: String, start: Long, pageSize: Int): List<Solidity.Address> {
+        val addressPage = mutableListOf<Solidity.Address>()
+        for (i in start until pageSize) {
```

Comment:
What if I want to load the second page starting at pagesize+1? 
Shouldn't this be like:
```
for (i in start until start + pageSize) {
[...]
}
```

Reply:
until is not inclusive